File: app/src/main/java/eu/kanade/mangafeed/ui/manga/myanimelist/MyAnimeListFragment.java
package eu.kanade.mangafeed.ui.manga.myanimelist;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.NumberPicker;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import java.text.DecimalFormat;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import eu.kanade.mangafeed.R;
import eu.kanade.mangafeed.data.database.models.MangaSync;
import eu.kanade.mangafeed.ui.base.fragment.BaseRxFragment;
import nucleus.factory.RequiresPresenter;

@RequiresPresenter(MyAnimeListPresenter.class)
public class MyAnimeListFragment extends BaseRxFragment<MyAnimeListPresenter> {

    @Bind(R.id.myanimelist_title) TextView title;
    @Bind(R.id.myanimelist_chapters) TextView chapters;
    @Bind(R.id.myanimelist_score) TextView score;
    @Bind(R.id.myanimelist_status) TextView status;

    private MyAnimeListDialogFragment dialog;

    private DecimalFormat decimalFormat = new DecimalFormat("#.##");

    public static MyAnimeListFragment newInstance() {
        return new MyAnimeListFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_myanimelist, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    public void setMangaSync(MangaSync mangaSync) {
        if (mangaSync != null) {
            title.setText(mangaSync.title);
            chapters.setText(mangaSync.last_chapter_read + "");
            score.setText(mangaSync.score == 0 ? "-" : decimalFormat.format(mangaSync.score));
            status.setText(getPresenter().myAnimeList.getStatus(mangaSync.status));
        }
    }

    private void showSearchDialog() {
        if (dialog == null)
            dialog = MyAnimeListDialogFragment.newInstance(this);

        dialog.show(getActivity().getSupportFragmentManager(), "search");
    }

    public void onSearchResults(List<MangaSync> results) {
        if (dialog != null)
            dialog.setResults(results);
    }

    @OnClick(R.id.myanimelist_title_layout)
    void onTitleClick() {
        showSearchDialog();
    }

    @OnClick(R.id.myanimelist_status_layout)
    void onStatusClick() {
        if (getPresenter().mangaSync == null)
            return;

        Context ctx = getActivity();
        new MaterialDialog.Builder(ctx)
                .title(R.string.status)
                .items(getPresenter().getAllStatus(ctx))
                .itemsCallbackSingleChoice(getPresenter().getIndexFromStatus(),
                        (materialDialog, view, i, charSequence) -> {
                            getPresenter().setStatus(i);
                            status.setText("...");
                            return true;
                        })
                .show();
    }

    @OnClick(R.id.myanimelist_chapters_layout)
    void onChaptersClick() {
        if (getPresenter().mangaSync == null)
            return;

        MaterialDialog dialog = new MaterialDialog.Builder(getActivity())
                .title(R.string.chapters)
                .customView(R.layout.dialog_myanimelist_chapters, false)
                .positiveText(R.string.button_ok)
                .negativeText(R.string.button_cancel)
                .onPositive((materialDialog, dialogAction) -> {
                    View view = materialDialog.getCustomView();
                    if (view != null) {
                        NumberPicker np = (NumberPicker) view.findViewById(R.id.chapters_picker);
                        getPresenter().setLastChapterRead(np.getValue());
                        chapters.setText("...");
                    }
                })
                .show();

        View view = dialog.getCustomView();
        if (view != null) {
            NumberPicker np  = (NumberPicker) view.findViewById(R.id.chapters_picker);
            // Set initial value
            np.setValue(getPresenter().mangaSync.last_chapter_read);
            // Don't allow to go from 0 to 9999
            np.setWrapSelectorWheel(false);
        }
    }

    @OnClick(R.id.myanimelist_score_layout)
    void onScoreClick() {
        if (getPresenter().mangaSync == null)
            return;

        MaterialDialog dialog = new MaterialDialog.Builder(getActivity())
                .title(R.string.score)
                .customView(R.layout.dialog_myanimelist_score, false)
                .positiveText(R.string.button_ok)
                .negativeText(R.string.button_cancel)
                .onPositive((materialDialog, dialogAction) -> {
                    View view = materialDialog.getCustomView();
                    if (view != null) {
                        NumberPicker np = (NumberPicker) view.findViewById(R.id.score_picker);
                        getPresenter().setScore(np.getValue());
                        score.setText("...");
                    }
                })
                .show();

        View view = dialog.getCustomView();
        if (view != null) {
            NumberPicker np  = (NumberPicker) view.findViewById(R.id.score_picker);
            // Set initial value
            np.setValue((int) getPresenter().mangaSync.score);
        }
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/myanimelist/MyAnimeListFragment.kt
package eu.kanade.tachiyomi.ui.manga.myanimelist

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.NumberPicker
import com.afollestad.materialdialogs.MaterialDialog
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.database.models.MangaSync
import eu.kanade.tachiyomi.ui.base.fragment.BaseRxFragment
import eu.kanade.tachiyomi.util.toast
import kotlinx.android.synthetic.main.card_myanimelist_personal.*
import kotlinx.android.synthetic.main.fragment_myanimelist.*
import nucleus.factory.RequiresPresenter
import java.text.DecimalFormat

@RequiresPresenter(MyAnimeListPresenter::class)
class MyAnimeListFragment : BaseRxFragment<MyAnimeListPresenter>() {

    companion object {
        fun newInstance(): MyAnimeListFragment {
            return MyAnimeListFragment()
        }
    }

    private var dialog: MyAnimeListDialogFragment? = null

    private val decimalFormat = DecimalFormat("#.##")

    private val SEARCH_FRAGMENT_TAG = "mal_search"

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_myanimelist, container, false)
    }

    override fun onViewCreated(view: View, savedState: Bundle?) {
        swipe_refresh.isEnabled = false
        swipe_refresh.setOnRefreshListener { presenter.refresh() }
        myanimelist_title_layout.setOnClickListener { onTitleClick() }
        myanimelist_status_layout.setOnClickListener { onStatusClick() }
        myanimelist_chapters_layout.setOnClickListener { onChaptersClick() }
        myanimelist_score_layout.setOnClickListener { onScoreClick() }
    }

    @Suppress("DEPRECATION")
    fun setMangaSync(mangaSync: MangaSync?) {
        swipe_refresh.isEnabled = mangaSync != null
        mangaSync?.let {
            myanimelist_title.setTextAppearance(context, R.style.TextAppearance_Regular_Body1_Secondary)
            myanimelist_title.setAllCaps(false)
            myanimelist_title.text = it.title
            myanimelist_chapters.text = if (it.total_chapters > 0)
                "${it.last_chapter_read}/${it.total_chapters}" else "${it.last_chapter_read}/-"
            myanimelist_score.text = if (it.score == 0f) "-" else decimalFormat.format(it.score)
            myanimelist_status.text = presenter.myAnimeList.getStatus(it.status)
        } ?: run {
            myanimelist_title.setTextAppearance(context, R.style.TextAppearance_Medium_Button)
            myanimelist_title.setText(R.string.action_edit)
            myanimelist_chapters.text = ""
            myanimelist_score.text = ""
            myanimelist_status.text = ""
        }

    }

    fun onRefreshDone() {
        swipe_refresh.isRefreshing = false
    }

    fun onRefreshError(error: Throwable) {
        swipe_refresh.isRefreshing = false
        context.toast(error.message)
    }

    fun setSearchResults(results: List<MangaSync>) {
        findSearchFragmentIfNeeded()

        dialog?.onSearchResults(results)
    }

    fun setSearchResultsError(error: Throwable) {
        findSearchFragmentIfNeeded()
        context.toast(error.message)

        dialog?.onSearchResultsError()
    }

    private fun findSearchFragmentIfNeeded() {
        if (dialog == null) {
            dialog = childFragmentManager.findFragmentByTag(SEARCH_FRAGMENT_TAG) as MyAnimeListDialogFragment
        }
    }

    fun onTitleClick() {
        if (dialog == null) {
            dialog = MyAnimeListDialogFragment.newInstance()
        }

        presenter.restartSearch()
        dialog?.show(childFragmentManager, SEARCH_FRAGMENT_TAG)
    }

    fun onStatusClick() {
        if (presenter.mangaSync == null)
            return

        MaterialDialog.Builder(activity)
                .title(R.string.status)
                .items(presenter.getAllStatus())
                .itemsCallbackSingleChoice(presenter.getIndexFromStatus(), { dialog, view, i, charSequence ->
                    presenter.setStatus(i)
                    myanimelist_status.text = "..."
                    true
                })
                .show()
    }

    fun onChaptersClick() {
        if (presenter.mangaSync == null)
            return

        val dialog = MaterialDialog.Builder(activity)
                .title(R.string.chapters)
                .customView(R.layout.dialog_myanimelist_chapters, false)
                .positiveText(android.R.string.ok)
                .negativeText(android.R.string.cancel)
                .onPositive { d, action ->
                    val view = d.customView
                    if (view != null) {
                        val np = view.findViewById(R.id.chapters_picker) as NumberPicker
                        np.clearFocus()
                        presenter.setLastChapterRead(np.value)
                        myanimelist_chapters.text = "..."
                    }
                }
                .show()

        val view = dialog.customView
        if (view != null) {
            val np = view.findViewById(R.id.chapters_picker) as NumberPicker
            // Set initial value
            np.value = presenter.mangaSync!!.last_chapter_read
            // Don't allow to go from 0 to 9999
            np.wrapSelectorWheel = false
        }
    }

    fun onScoreClick() {
        if (presenter.mangaSync == null)
            return

        val dialog = MaterialDialog.Builder(activity)
                .title(R.string.score)
                .customView(R.layout.dialog_myanimelist_score, false)
                .positiveText(android.R.string.ok)
                .negativeText(android.R.string.cancel)
                .onPositive { d, action ->
                    val view = d.customView
                    if (view != null) {
                        val np = view.findViewById(R.id.score_picker) as NumberPicker
                        np.clearFocus()
                        presenter.setScore(np.value)
                        myanimelist_score.text = "..."
                    }
                }
                .show()

        val view = dialog.customView
        if (view != null) {
            val np = view.findViewById(R.id.score_picker) as NumberPicker
            // Set initial value
            np.value = presenter.mangaSync!!.score.toInt()
        }
    }

}
